server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -839,19 +844,30 @@ public void write(XContentBuilder b) throws IOException {
                 b.startObject(simpleName());
             }
 
-            for (SourceLoader.SyntheticFieldLoader field : fields) {
-                if (field.hasValue()) {
-                    field.write(b);
+            if (ignoredValues != null && ignoredValues.isEmpty() == false) {
+                Set<String> storedFields = new HashSet<>(ignoredValues.size());
+                for (var value : ignoredValues) {
+                    storedFields.add(value.name());
+                }
+                for (SourceLoader.SyntheticFieldLoader field : fields) {
+                    // Check if the field source is stored separately, to avoid double-printing.
+                    if (field.hasValue() && storedFields.contains(field.fieldName()) == false) {
+                        field.write(b);
+                    }
                 }
-            }
-            hasValue = false;
-            if (ignoredValues != null) {
                 for (IgnoredSourceFieldMapper.NameValue ignored : ignoredValues) {
                     b.field(ignored.getFieldName());
```

Comment:
I didn't think about this before - doesn't this break the alphabetic sorting of fields in the source? We say in documentation that it is sorted.

Reply:
This is a good point. I'll work on a follow-up PR to restore ordering.

Reply:
Well, I gave it a try here. While it's easy to order fields at the object level, we don't do so for the contents of the object array that get stored and retrieved verbatim. I added tests for this, not sure if there's an easy way to fix this..

Reply:
I think that's exactly what we need, we only specify ordering of fields.